### ISSUE TITLE
API Retrievers should return Objects instead of text

### ISSUE THREAD

**Issue by PonteIneptique** [Capitains/MyCapytain#89]
Everyone assigned is welcome to give some feedback.

Right now, CTS API retrievers (There is two implementation : one in MyCapytain, one in Nautilus) returns the data/bytes from the API request, which means that right now, when we use Nautilus, we produce a huge overhead when used in connection with Nemo. Here is what happens
1. Nemo make a GetPassage request
2. Nemo request for information as http (text, reffs or inventory)
3. Nautilus find the informations (= multiple steps not shown here)
4. Nautilus compiles the informations to XML
5. Nautilus encodes it as string
6. Nautilus produce a http answer
7. Nemo process the http answers
8. Nemo parse the XML in the http answers through lxml
9. Nemo parse the XML Object (Lxml object) as a Passage.

You can see that there is at least 2, 5, 6, 7 and 8 that consumes a large amount of memory AND process and that are not necessary. There is half of the steps that are used that should not.

As we want the best for everyone, I was thinking that we should start to change the current behavior of `MyCapytains.retrievers.proto.*` which should by default return resources that are understandable by said Python libraries (related to #88 more or less) : 
- Passage
- InventoryRessource (GetLabel and GetCapabilities)
- URNSet : First URN, PrevNext URN, Children URNs

It means shifting from the current "paradigm" of MyCapytain ( Object parsing retrievers ) to a more "closed one" (Retrievers returning Object). The way to implement would probably to refactor current "retriever" module as a "requester" module and create a real retriever module which could be a little more decoupled from CTS given the fact DTS is around the corner.

Thoughts ?

**Comment by PonteIneptique** [comment]
As a set of function, retrievers could have : 

- getMetadata(filters) -> InventoryResource
- getText(id, PassageReference) -> Passage
- getPassageSiblings(id, PassageReference) -> ([PassageReference], [PassageReference])
- getPassageChildren(id, PassageReference, level) -> [PassageReference]

**Comment by fbaumgardt** [comment]
This sounds good to me. I would suggest setting up little bit of a class hierarchy for retrievers, and provide a basic retriever which returns bytes or strings. You can extend it to create the object based retriever, while keeping the lower level interface for cases where it's helpful.

**Comment by PonteIneptique** [comment]
So basically something along these lines ?

```
Retriever
    |- BytesRetriever
    |- ObjectRetriever
        |- CTSObjectRetriever 
```

I think having requesters vs. Retriever might be a clearer for people ?
Aka : 
```
Resolver
    |- Requester
    |- Retriever
        |- CTSObjectRetriever 
```
or simply
```
Requester
     |- CTSRequester
     |- DTSRequester
Retriever
     |- CTSObjectRetriever 
```
?

**Comment by fbaumgardt** [comment]
I thought of the retriever as a subclass of requester. I.e. you'd call super and then parse the bytes/string. However -- I don't know enough about inheritance I python and how resource usage is in this case, I.e. if inheriting and calling super is consuming the same resources as your current architecture or less.

**Comment by PonteIneptique** [comment]
I see. I'll see what I can do with this.

**Comment by balmas** [comment]
I'm not sure I fully understand the distinction you're making between requester and retriever.

**Comment by PonteIneptique** [comment]
Fixed with 2.0.0b0 in #95 / #90 : MyCapytain.resolvers exists aside of retrievers
